Add router

johnpooch opened the pull request:
Adds a `Router` component to the component hierarchy to support basic routing for pages like About, Settings, etc. For now I've only added About to just test this.

Also adds a few other bits of infrastructure to the project:
* Adds typescript support. About.tsx, App.tsx, and Router.tsx, as well as tests scripts are typescript.
* Fixes some type problems with About (fontWeight takes an integer and fontColor should be color)
* Adds tests
* Adds tests to github CI (see below)
* Moves what was App into a file called LegacyApp (sorry I couldn't think of a better name). This is a temporary thing.

**Testing**
I'm not sure if our ActivityContainer thing is something that we need to make the app work in the android wrapper. I'd like to test that normal usage or react router works in the android wrapper context.

zond commented on web-app/src/components/ChatMenu.js:
```diff
@@ -64,28 +66,14 @@ export default class ChatMenu extends React.Component {
+									const activeChannel = js.Properties.find(
+										(c) => c.Properties.Members.join(",") === decodedChannelId
+									);
+									this.setState({ activeChannel })
```
Should you really do a `this.setState` within a `this.setState`?

zond replied:
I fixed this :D

zond commented on web-app/src/components/Game.js:
```diff
@@ -62,6 +64,21 @@ export default class Game extends React.Component {
+			.then((resp) => resp.json());
+		}
+
+		this.close = () => {
```
How come this is an anon func defined in the constructor and not a regular class method?

zond replied:
I changed it.

zond commented on web-app/src/pages/Router.tsx:
```diff
@@ -0,0 +1,50 @@
+export const RouteConfig = {
+	About: '/about',
+	Game: '/game/:gameId',
+	GameChatChannel: '/game/:gameId/Channel/:channelId/Messages',
```
If the "game" path element is in lowercase, shouldn't the "channel" and "messages" elements be as well?

zond replied:
And "lab" below.

zond replied:
I changed it.

zond commented on web-app/src/pages/Router.tsx:
```diff
@@ -0,0 +1,50 @@
+	About: '/about',
+	Game: '/game/:gameId',
+	GameChatChannel: '/game/:gameId/Channel/:channelId/Messages',
+	GameLabratoryMode: '/game/:gameId/Lab/:labOptions',
```
s/GameLabratoryMode/GameLaboratoryMode/ ?

zond replied:
I fixed it.

zond commented on web-app/src/pages/Router.tsx:
```diff
@@ -0,0 +1,50 @@
+			<Route exact path={RouteConfig.GameChatChannel}>
+				<Game chatOpen />
+			</Route>
+			<Route exact path={RouteConfig.GameLabratoryMode}>
```
s/GameLabratoryMode/GameLaboratoryMode/ and s/labratoryMode/laboratoryMode/ right?

johnpooch replied:
Doesn't make much difference :) Whichever you prefer. I included the game prefix because it vaguely reflects the url itself but it really doesn't matter :)

zond commented:
Nice work!

The router really simplifies things :D

I fixed a lot of issues in this branch caused by deprecation (I evidently managed to get new node stuff installed), some issues I found and commented about, and the lab mode.

The remaining issue - that's nontrivial to fix - is that I used to capture popstate events to close dialogs with the back button (which works well in Android as well), but the react router also captures those events, and I haven't yet figured out a good way to fix that.

In `index.js` is an event listener for popstate, and in `helpers.js` is the functions `onback` and `genOnback` that hook callables that get run (and removed) when a popstate happens. This is used to close dialogs.

The dialogs also have a `close` function that generally uses another helper function that removes their `onback` function so that the back queue gets cleaned up.

One good way to solve this would be to give all dialogs a route, but that might be overkill? Especially if we want to get rid of the dialogs.

Can you think of a quick hack that makes the current dialogs work (with "back"-closing) until we figure out what to do with them?

johnpooch commented:
> Nice work!
> 
> The router really simplifies things :D
> 
> I fixed a lot of issues in this branch caused by deprecation (I evidently managed to get new node stuff installed), some issues I found and commented about, and the lab mode.
> 
> The remaining issue - that's nontrivial to fix - is that I used to capture popstate events to close dialogs with the back button (which works well in Android as well), but the react router also captures those events, and I haven't yet figured out a good way to fix that.
> 
> In `index.js` is an event listener for popstate, and in `helpers.js` is the functions `onback` and `genOnback` that hook callables that get run (and removed) when a popstate happens. This is used to close dialogs.
> 
> The dialogs also have a `close` function that generally uses another helper function that removes their `onback` function so that the back queue gets cleaned up.
> 
> One good way to solve this would be to give all dialogs a route, but that might be overkill? Especially if we want to get rid of the dialogs.
> 
> Can you think of a quick hack that makes the current dialogs work (with "back"-closing) until we figure out what to do with them?

Okay I'll take a look at this. I'm not sure I fully understand what you are describing here but my initial thought is this.

In general we should use querystring parameters to control whether dialogs are open or closed. This allows us to have a link directly to a page with the dialog open which is nice. So when we hit the close button we should simply remove the querystring param from the url and this should cause the UI to update and remove the dialog.

I would have used querystring params for chat but it already uses url params and I didn't want to break any exisiting links.

I think we should use url params for pages and querystring params for ui elements on pages. As mentioned, we should probably not be using dialogs as much as we are so we can transition querystring params to url params as we go.

zond commented:
Alright.

I kind of feel that down to the level of chat channels path elements are nice, but dialogs (and here I mean real dialogs - temporary floating boxes) should be query params.

I'm experimenting a bit with using a query param for a dialog now. Let's see how easy the transition is...

zond commented:
> Hey yeah good questions. I think that it should be possible to use a qsparam to control the nation preferences dialog in all of these situations. My advice would be to make the param value be the id of the game in that case of the browse open games page, e.g. `?nationPreferencesDialogOpen=1234` and in the context of the create game and game pages the param could just be `?nationPreferencesDialogOpen=1`.

Of course that's fine in the join game-case, but for the create game case it doesn't make sense to even show the dialog when you load the URL. What we _should_ do in that case I guess is encode the entire state of the app in the URL (which would be sort of correct to do in any case all over the app to make any app state you view loadable from the URL).

So, either we accept that we load a bogus nation selector for a create game view (that will just create and join a standard game when the URL is reloaded), or we don't use a URL for the nation selector in the create game view, or we make sure the entire create game view state is in the URL... :O

> On a related note, it looks like there is too much state in the NationPreferences component. The only state it needs to track is the order of the nations, e.g. `[{ England: 2, France: 3, Germany: 1... }]`, The onSelected function should be a prop and should not be assigned to state. It shouldn't have an open state but should receive an open prop from the parent element.

Yes that was one of the things I was working on when I ran into the above mentioned problems ... :/

> I wouldn't think of it as having one global NationPreferences element but instead having one for each page. On the browse game page there should be only one nation preferences element and it's nations would be determined by a prop which is in turn determined by the qs param id

As above.

> I don't think any of these are meaningless to load except for the create game one. If the form hasn't been set to "Preference" then the dialog shouldn't appear. But once the qs is present and the "Preference" option has been set in the form, the dialog should appear

I agree (with the create game view as a caveat).

johnpooch commented:
> @johnpooch Alright, this seems to work. I still don't like it that we get a "pointless" nation selector if the page is reloaded during game creation, but the rest is pretty neat, and that problem isn't important enough to block the rest.
> 
> I suggest we fix back button for all dialogs in this branch this way, and when it's done we merge it into the beta?
> 
> Hopefully we can merge beta into master before that, but I'm slightly worried about the reports of the map scrolling problems - we didn't change that in the new code, so why does it act up? :O

Nice one :) Will take a proper dig through this later and drops some comments. Regarding game create page, encoding the state in url sounds like a kinda messy approach. For now what you're suggesting makes sense. Can't see why anyone would want to use that url anyways. 

Moving forward it might be sensible to separate the two steps. After creating a game the user is redirected to the join game page. That way the game entity exists in all cases when using nation preferences component. Obvs slight issue there is what happens when the user doesn't bother adding preferences. Then they wouldn't be added as a participant and we'd have games with no participants. A problem for a later day :) But overrall I think it's best to separate out these steps where possible. It makes the application logic more simple.

Btw, soon when we have fully embraced the beauty of functional components and hooks something like this would a nice api:
```
const { value, add, remove } = useQuerystringParam('nationPreferencesDialogOpen');
```
Where value is the current value and add and remove are functions which can be called to add or remove the param from search

zond commented:
> Nice one :) Will take a proper dig through this later and drops some comments. Regarding game create page, encoding the state in url sounds like a kinda messy approach. For now what you're suggesting makes sense. Can't see why anyone would want to use that url anyways.

Fair enough.

> Moving forward it might be sensible to separate the two steps. After creating a game the user is redirected to the join game page. That way the game entity exists in all cases when using nation preferences component. Obvs slight issue there is what happens when the user doesn't bother adding preferences. Then they wouldn't be added as a participant and we'd have games with no participants. A problem for a later day :) But overrall I think it's best to separate out these steps where possible. It makes the application logic more simple.

Open games without players don't exist, and I'd prefer it if that stayed true :o

> Btw, soon when we have fully embraced the beauty of functional components and hooks something like this would a nice api:
> 
> ```
> const { value, add, remove } = useQuerystringParam('nationPreferencesDialogOpen');
> ```
> 
> Where value is the current value and add and remove are functions which can be called to add or remove the param from search

Yeah, that'd look nice :)

johnpooch commented:
Hey @zond added some functionality to lift the stats dialog out of lower level components. Now it lives in a wrapper and the functionality to open a dialog is provided to sub-components through context. It seems to be working pretty nicely. Just need to add it to NationAvatar

zond commented:
> Hey @zond added some functionality to lift the stats dialog out of lower level components. Now it lives in a wrapper and the functionality to open a dialog is provided to sub-components through context. It seems to be working pretty nicely. Just need to add it to NationAvatar

Do I understand correctly that this makes us have exactly one single stats dialog, and every time it's opened it will just activate that one?

What's the benefits and drawbacks of that solution compared to the one used for nation selection (having one dialog for every usage of the nation selection)?

johnpooch commented:
> > Hey @zond added some functionality to lift the stats dialog out of lower level components. Now it lives in a wrapper and the functionality to open a dialog is provided to sub-components through context. It seems to be working pretty nicely. Just need to add it to NationAvatar
> 
> Do I understand correctly that this makes us have exactly one single stats dialog, and every time it's opened it will just activate that one?
> 
> What's the benefits and drawbacks of that solution compared to the one used for nation selection (having one dialog for every usage of the nation selection)?

Yep that's pretty much how it works. The benefit is that it centralizes the logic for how the dialog is managed. This means that lower level components don't need to have the stats dialog as a sub component. It allows for simpler lower level components. For example, the NationAvatar component shouldn't have StatsDialog as a sub component. It should just have an onClick function. It makes NationAvatar more flexible, easier to test, and easier to reason about.

Similarly, by having the StatsDialog logic in a wrapper we can define it's behaviour in a single place. This makes it easier to maintain, test, reason about, etc.

zond commented:
> > Do I understand correctly that this makes us have exactly one single stats dialog, and every time it's opened it will just activate that one?
> > What's the benefits and drawbacks of that solution compared to the one used for nation selection (having one dialog for every usage of the nation selection)?
> 
> Yep that's pretty much how it works. The benefit is that it centralizes the logic for how the dialog is managed. This means that lower level components don't need to have the stats dialog as a sub component. It allows for simpler lower level components. For example, the NationAvatar component shouldn't have StatsDialog as a sub component. It should just have an onClick function. It makes NationAvatar more flexible, easier to test, and easier to reason about.
> 
> Similarly, by having the StatsDialog logic in a wrapper we can define it's behaviour in a single place. This makes it easier to maintain, test, reason about, etc.

So why not solve the nation selector dialog the same way?

johnpooch commented:
> > > Do I understand correctly that this makes us have exactly one single stats dialog, and every time it's opened it will just activate that one?
> > > What's the benefits and drawbacks of that solution compared to the one used for nation selection (having one dialog for every usage of the nation selection)?
> > 
> > 
> > Yep that's pretty much how it works. The benefit is that it centralizes the logic for how the dialog is managed. This means that lower level components don't need to have the stats dialog as a sub component. It allows for simpler lower level components. For example, the NationAvatar component shouldn't have StatsDialog as a sub component. It should just have an onClick function. It makes NationAvatar more flexible, easier to test, and easier to reason about.
> > Similarly, by having the StatsDialog logic in a wrapper we can define it's behaviour in a single place. This makes it easier to maintain, test, reason about, etc.
> 
> So why not solve the nation selector dialog the same way?

We should! :)

zond commented:
> We should! :)

Great :D
